marconi-core/src/Marconi/Core/TracedStorable.hs

```diff
@@ -0,0 +1,391 @@
+    CNRollForward !pt
+  | CNRollBack    !pt
+ ---- | CNApplication !(StorableNotifications h)
+  deriving (Generic)
+
+type family StorableMonad h :: * -> *
+
+{-
+   The first, `Buffered` class explains what it means for an indexer to be accumulating
+   a set of results that will be flushed to disk when the memory buffer is fully filled.
+
+   The memory layout of the indexer has been simplified by a lot since the previous
+   version and it looks something like this:
+
+   |e|e|e|e|e|e|e|e|e|e|e|e|e|e|e|e|e|e|                                     |
+   |-------------------|---------------|-------------------------------------|
+   |   memory/buffer   |  disk/events  |       disk/aggregate (optional)     |
+
+   When the buffer is filled with events, they all get flushed to disk. We need this
+   operation for performance reasons. Most databases will have significantly improved
+   performance for batch inserts.
+
+   Rollbacks will only happen in the memory or disk/events storage area, so the
+   developer has to make sure that the allocated space for the buffer and the
+   disk/events is greater than the K parameter.
+
+   The last part disk/aggregate represents some aggregated data. At some point the
+   developer should write code to aggregate the disk/events into the aggregate
+   section of the database. We did not provide any special handling of this for the
+   following reasons:
+
+     1) A lot of indexers will not require aggregated data. Only storing events should
+        be enough for most applications.
+     2) Data aggregation can be implemented at database level and probably be more
+        efficient than what Haskell can do here.
+     3) We can always extend the interface to include direct support for this pattern
+        if there is a demand from the users of the API.
+-}
+class Buffered h where
+  -- | This function persists the memory/buffer events to disk when the memory buffer is filled.
+  persistToStorage :: Foldable f => f (StorableEvent h) -> h -> StorableMonad h h
+
+  {- This function retrieves the events from the disk/events area.
+     If the user chooses to only store events, without accumulating them, this function
+     is expected to return the events over which rollbacks can occur in order to keep
+     things performant -}
+  getStoredEvents :: h -> StorableMonad h [StorableEvent h]
+
+{-
+   All information from indexers should be made accessible through queries. Queries
+   act a lot like folds over the event stream. Each query introduces two indexed types:
+   one for the type of requests (called StorableQuery) and one for responses (called
+   StorableResult).
+
+   All queries include a validity interval. If the data is not available for a specified
+   interval, the returned result should specify that. It is also recommended that the result
+   includes the slot number at which the query was ran.
+-}
+class Queryable h where
+  queryStorage
+    :: Foldable f
+    => StorablePoint h
```

Comment:
Why not `Maybe (StorablePoint h)` here, to allow a simple "give me the freshest thing you have" that may be suitable in some scenarios.

Reply:
Nice observation. I debated that myself and @kayvank. I don't think that give me the freshest you have is something we want to support. As you further notice, that would require changing the result type to include an indication of the slot number, but that is inherently unsafe (what if some user just ignores the information?).

Always knowing the last processed block is one of the features allowed by the notification system, so the user will be able to always know what is the latest block, so they can specify it in the query.